src/util.rs

```diff
@@ -1,23 +1,91 @@
 use std::path::Path;
 use std::{fs::File, io::Read};
 
-use anyhow::Context;
+use anyhow::{bail, Context};
 use rustdoc_types::Crate;
+use serde::Deserialize;
 
 pub(crate) const SCOPE: &str = "semver-checks";
 
+#[derive(Deserialize)]
+struct RustdocFormatVersion {
+    format_version: u32,
+}
+
 pub(crate) fn load_rustdoc_from_file(path: &Path) -> anyhow::Result<Crate> {
     // Parsing JSON after fully reading a file into memory is much faster than
     // parsing directly from a file, even if buffered:
     // https://github.com/serde-rs/json/issues/160
     let mut s = String::new();
     File::open(path)
-        .with_context(|| format!("Failed to open rustdoc JSON output file {:?}", path))?
+        .with_context(|| format!("failed to open rustdoc JSON file {}", path.display()))?
         .read_to_string(&mut s)
-        .with_context(|| format!("Failed to read rustdoc JSON output file {:?}", path))?;
+        .with_context(|| format!("failed to read rustdoc JSON file {}", path.display()))?;
+
+    match serde_json::from_str(&s) {
+        Ok(value) => Ok(value),
+        Err(e) => {
+            // Attempt to figure out the more precise reason the deserialization failed.
+            // Several possible options and their resolutions:
+            // (1) The file isn't actually a rustdoc JSON file. The user should supply a valid file.
+            // (2) The rustdoc JSON file has a version number that is too old, and isn't supported.
+            //     The user should upgrade to a newer nightly Rust version and regenerate the file.
+            // (3) The rustdoc JSON file has a version number that is too new, and isn't supported.
+            //     The user should attempt to upgrade to a newer cargo-semver-checks version
+            //     if one is already available, or open a GitHub issue otherwise.
+
+            // The error on this line is case (1).
+            let version = serde_json::from_str::<RustdocFormatVersion>(&s).with_context(|| {
+                format!(
+                    "parsing rustdoc JSON failed, and could not determine rustdoc format version \
```

Comment:
nit: Personally, I would leave out "and could not determine rustdoc format version", that is an implementation detail that the user can't act upon.

As hybrid, you could instead make it "Unrecognized rustdoc output" or something